Automattic/wp-calypso
# DRAFT | Launchpad: Integrate with Real Data from Checklist Endpoint

# Proposed Changes
This PR is the FE integration of the newly created `launchpad/checklist` endpoint which provides the launchpad checklist. Up until now, the launchpad tasks data has been defined entirely on the FE. This PR removes all unnecessary mutations of the data between the time of fetching the data and displaying it to the user.

# Testing Instructions
This PR touches a lot of code in different places, so the testing can be broken into three different parts:

1. Testing the redirect from Launchpad to My Home when Launchpad is completed
2. Testing the redirect from My Home to Launchpad when Launchpad is still enabled
3. Testing the functionality of all Launchpad tasks

## Testing Setup
1. Checkout this branch

2. Run `yarn` and `yarn start` if needed

3. Sync a recent Jetpack branch that contains the latest changes for the `launchpad/checklist` endpoint. You can use [fix/launchpad_link_in_bio_launch](https://github.com/Automattic/jetpack/pull/30176):
`bin/jetpack-downloader test jetpack-mu-wpcom-plugin fix/launchpad_link_in_bio_launch`
Alternatively, if that branch is already merged at the time of testing this you can sync to trunk:
`bin/jetpack-downloader test jetpack-mu-wpcom-plugin trunk`

4. Sandbox `public-api.wordpress.com` and make sure to sandbox the new testing site you're about to create

## Testing My Home -> Launchpad Redirect
1. Create a new Launchpad site and make sure to sandbox it
2. Once you arrive at the Launchpad screen, navigate to `http://calypso.localhost:3000/home/YOUR_SITE_SLUG`
3. Confirm you are redirected back to the Launchpad screen

## Testing Launchpad -> My Home Redirect
1. Create a new Launchpad site and make sure to sandbox it
2. Once you arrive at the Launchpad screen, copy the URL
3. Complete all Launchpad tasks and launch the site
4. You should get redirected to My Home
5. Manually go back to the URL you have copied in Step 2
6. Confirm you are redirected back to My Home

Note: You would see the Launchpad screen briefly before getting redirected, I think this is possibly a follow-up on how we handle data caching / refetching...

## Testing Launchpad Tasks
Now the fun part... We need to test all tasks for all flows to make sure everything behaves as expected.

### Write Flow
1. Go to http://calypso.localhost:3000/start to create a new site
2. Once you arrive at the Goals page, select Write & Publish and click Continue
3. Once you arrive at Launchpad, open the same site on `wordress.com` and start comparing the behavior side to side
4. Confirm all tasks work as expected and you are able to launch your site

### Build Flow
1. Go to http://calypso.localhost:3000/start to create a new site
2. Once you arrive at the Goals page, don't select anything and click Continue
3. Once you arrive at Launchpad, open the same site on `wordress.com` and start comparing the behavior side to side
4. Confirm all tasks work as expected and you are able to launch your site

### Link in Bio Flow
1. Go to http://calypso.localhost:3000/setup/link-in-bio/intro to create a new site
2. Once you arrive at Launchpad, open the same site on `wordress.com` and start comparing the behavior side to side
3. Confirm all tasks work as expected and you are able to launch your site

### Newsletter Flow
1. Go to http://calypso.localhost:3000/setup/newsletter/intro to create a new site
2. Once you arrive at Launchpad, open the same site on `wordress.com` and start comparing the behavior side to side
3. Confirm all tasks work as expected and you are able to launch your site

### Free Flow
1. Go to http://calypso.localhost:3000/setup/free/intro to create a new site
2. Once you arrive at Launchpad, open the same site on `wordress.com` and start comparing the behavior side to side
3. Confirm all tasks work as expected and you are able to launch your site

### VideoPress Flow
1. Go to http://calypso.localhost:3000/setup/videopress/intro to create a new site
2. Once you arrive at Launchpad, open the same site on `wordress.com` and start comparing the behavior side to side
3. Confirm all tasks work as expected and you are able to launch your site
